### PR TITLE
Model and Backend Endpoints

### PR DESCRIPTION
**Pull request by heliobmartins**
### Goals of the Pull Request:
- How does the project fit under Python Guidelines?
- Does the API's are following good RESTFul style?
- Thoughts about Project Structure?
- Any insight, please, feel free to comment or amend.
- There is no **update/delete** yet. (You can ignore it for now). As soon as I finish it, I will send you a new PR if you don't mind.

**Context of the repository:**
- Learn a new programming language;
- Proof of Concept to test different architectures;
- Understand the limitations of the platform;

_I intend to create more proof of concepts following this idea._

### CODE REVIEW

File: services/institution_service.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        institutions, next_cursor, more = query.fetch_page(page_size=request.page_size, start_cursor=cursor)
+        response = InstitutionListResponse()
+        response.institutions = [Institution.encode(i) for i in institutions]
+        if more and next_cursor:
+            response.cursor = next_cursor.urlsafe()
+        return response
+
+    def _validate_entity_creation(self, new_institution):
+        pass
+        # raise endpoints.BadRequestException("Error while creating Institution")
```
Comment (damncabbage) [inline services/institution_service.py:41]: I'm pretty sure this is a TODO item, but am flagging it just to make sure it's not forgotten.

File: services/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from services import *
```
Comment (damncabbage) [inline services/__init__.py]: I'm getting the impression this is boilerplate, as this one-liner (AFAICT) doesn't do anything.

For more detail, have a [read about `from module import *`](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package).

File: utils/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+import caas_utils
```
Comment (damncabbage) [inline utils/__init__.py]: Ditto this file. If you don't see any initialisation code being run in the `caas_utils`, then this file can be empty (the empty file just acting as a "you can import from this directory" marker).

File: messages/referral_converter.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+                            name=referral.name,
+                            email=referral.email,
+                            phone=referral.phone,
+                            consent=referral.consent,
+                            institution=Institution.encode(referral.institution_id.get()))
+
+
+def convert_to_into_entity(referral_to):
+    # @Marcot, this could be a problem as there are cross reference.
+    from domain.referral_domain import Referral
```
Comment (damncabbage) [inline messages/referral_converter.py:18]: Uh, it seems weird to do an import in the middle of a function. I didn't think you could actually do this.

**Comment by damncabbage** [comment]
Just to preface this: I'm noticing some familiar concepts. I'm treating my commentary as an unrelated side-project that uses these concepts because it's a useful cohesive domain to model. Please don't consider this a code review for the sake of any internal CancerAid signoff, or me contributing to any CancerAid business processes.

---

General comments: looks fine. 🙂 Some notes.

> feel free to comment

Before we get to the more specific stuff, I had a (non-Python) question:
- How does `caasv1openapi.json` stay in sync with the implementation itself? At the moment it looks like it's manually encoded twice (in Python, and in JSON). If it's manual, how do you plan to check it to stop it drifting?

---

> How does the project fit under Python Guidelines?

As a general thing, I'd recommend using an autoformatter tool like [autopep8](https://pypi.org/project/autopep8/) to handle some of this for you, just so you don't have to remember the [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines from memory.

---

> Does the API's are following good RESTFul style?

"RESTful" is super-nebulous, for better or worse. For reference for this discussion, the Rails / Phoenix API style is roughly:

Things (collection):
- `GET /api/things` – list the existing Things,
- `POST /api/things` – create a new Thing,

Thing (single item):
- `GET /api/things/:id` – get a Thing identified by the string in `:id`,
- `PATCH /api/things/:id` – update/edit the Thing identified by the string in `:id`.
- `DELETE /api/things/:id` – delete the Thing identified by the string in `:id`.

There's argument that the Rails version here is "RESTful" because it's modelling in terms of operations (GET/POST/PATCH) on resources (eg. a 'Collection of Things' being one resource, and 'an Individual Thing' being another).

(Rails fucked this up by having its restful routes as before, but then tossing in UI/webpage-specific ones like `/new` and `/edit` such that the Original Philosophy doesn't seem to fit anymore.)

To a crowd familiar to the _"make a resource and define operations on it"_ philosophy (that Rails maybe started and people from other communities followed), you *may* get a strange reaction to the `/api/thing/list/<id>` and `/api/thing/new` routes, because it seems like you're using sub-routes (eg. `/new`) to define operations, instead of the HTTP GET/POST/PATCH/DELETE verbs.

… But opinion is all over the place on this one, so choose something you and anyone else you're working with is comfortable with and just go with it.

---

> Thoughts about Project Structure?

The good:

Segmentation, in this case by purpose, means you and others know how to extend this application. If it gets very big, though, you might decide to flip from "by-type" segregation (`services/`, `messages/`, etc) into "domain" segregation (`institutions/services/`, `institutions/messages/`, etc), but that's probably a ways off.

The less good:

You're getting into the weeds a bit with domain and messages modules cross-importing each other. The place where I see it happening is, for example, a class in domain having a method that uses Messages to do further encoding.

I'm an FP wonk so I'm biased, but I would try in this case to separate the domain classes defining "shape" from the domain class methods defining "transformation" (encoding / decoding). Keep the types (shapes) in domain, and keep the transformations of those shapes in messages.

(If this were a typed language, I'd be treat "domain" as internal-to-the-program types, with "messages" as functions knowing how to transform JSON to and from them.)

**Comment by damncabbage** [comment]
(I'm a little strapped for time at the moment; any further comments on my part will have to wait for later in the week when I get some free time again.)

File: services/__init__.py (hunk 2 above)
Comment (heliobmartins) [reply services/__init__.py]: This is extremelyyy, tricky! I will try to get used to that. Thought I had to have them defined in this file to compile. =O

Thanks Rob!

File: utils/__init__.py (hunk 3 above)
Comment (heliobmartins) [reply utils/__init__.py]: I think I understand it. This is quite tricky as I said previously.

File: messages/referral_converter.py (hunk 4 above)
Comment (heliobmartins) [reply messages/referral_converter.py:18]: Hahaha!

This is UGLY, but to make the code compile, this was the only way I could do as there is currently a "cross-dependency problem".

File: services/institution_service.py (hunk 1 above)
Comment (heliobmartins) [reply services/institution_service.py:41]: Yes, it's on my TODO list. I did not forget about it. Thanks anyway.

**Comment by heliomartins** [comment]
> f

**Comment by heliobmartins** [comment]
> (I'm a little strapped for time at the moment; any further comments on my part will have to wait for later in the week when I get some free time again.)

No worries ROB!
I know you are busy, I'm completely grateful for your help.

**Comment by heliobmartins** [comment]
@damncabbage 

This is completely aside from my current organization. This is a mere proof of concept.

In regards to your questions, here we go:

> How does caasv1openapi.json stay in sync with the implementation itself? At the moment it looks like it's manually encoded twice (in Python, and in JSON). If it's manual, how do you plan to check it to stop it drifting?`

I didn't understand much of your question... 🤪I will try to reply anyway, let me know if it's correct, please. 
I'm currently running the command manually, as you can see in the [README.md](https://github.com/heliomartins/appengine-python27-standard/blob/master/README.md). The main reason for that is: When I send a new version of this file to Google Cloud Endpoints, it generates a Developer Console Portal, where I can have access through all my APIS and Test them. To avoid the manual work and keep them update, I reckon would be necessary to create a `script (.sh)` that as soon as a new version is deploye successfully, this file is regenerated and automatically uploaded with the latest version to GCE. Below, an screenshot of the portal I'm talking about.

image


> "RESTful" is super-nebulous, for better or worse. For reference for this discussion, the Rails / Phoenix API style is roughly:

Yes, you're right once more. I'm actually trying to write the apis following [Ruby docs](https://guides.rubyonrails.org/routing.html#crud-verbs-and-actions)

Thanks once more @damncabbage

**Pull request merged**